File: backend/app/services/google_auth_service.py
import httpx
from fastapi import HTTPException

from app.core.config import settings


async def verify_google_credential(credential: str) -> dict:
    async with httpx.AsyncClient(timeout=10) as client:
        response = await client.get("https://oauth2.googleapis.com/tokeninfo", params={"id_token": credential})
    profile = response.json() if response.headers.get("content-type", "").startswith("application/json") else {}
    if response.status_code >= 400:
        raise HTTPException(status_code=401, detail=profile.get("error_description") or "Google sign-in verification failed.")
    if profile.get("aud") != settings.google_client_id:
        raise HTTPException(status_code=401, detail="Google credential audience does not match this app.")
    return profile

File: backend/src/services/googleAuthService.js
export async function verifyGoogleCredential(credential) {
  const params = new URLSearchParams({ id_token: credential });
  const googleResponse = await fetch(`https://oauth2.googleapis.com/tokeninfo?${params.toString()}`);
  const profile = await googleResponse.json().catch(() => ({}));
  if (!googleResponse.ok) throw Object.assign(new Error(profile.error_description || 'Google sign-in verification failed.'), { statusCode: 401 });
  if (profile.aud !== process.env.GOOGLE_CLIENT_ID) throw Object.assign(new Error('Google credential audience does not match this app.'), { statusCode: 401 });
  return profile;
}
